### ISSUE TITLE
Github action to automatically update precompile files

### ISSUE THREAD

**Issue by aminya** [timholy/SnoopCompile.jl#37]
Automatic repository updating using Github actions for uploading the files generated by snoopCompile.jl script

Similar to https://github.com/julia-actions/julia-uploadcodecov or https://github.com/julia-actions/julia-docdeploy

**Comment by timholy** [comment]
Cool idea. These seem to require JavaScript, a language I don't speak, so I'd rely on external contributors for such bots.

**Comment by aminya** [comment]
Also, I found this push bot: https://github.com/ad-m/github-push-action which does exactly what we want. We just need to run the Julia file before this.

**Comment by aminya** [comment]
Probably @davidanthoff can help us with this

**Comment by aminya** [comment]
I made a workflow file that does what we want! 😃 
https://github.com/juliamatlab/MatLang/blob/master/.github/workflows/SnoopCompile.yml

**Comment by davidanthoff** [comment]
Cool! Once you guys have tried it out for a while I'd be happy to create a proper github action and integrate it with https://github.com/davidanthoff/PkgButlerEngine.jl.

**Comment by aminya** [comment]
I get this error: https://github.com/aminya/AcuteML.jl/commit/141099db133d89c147f94779d06a20cdbed6be65/checks?check_suite_id=337167681#step:7:5
Do you know what the reason is?
```julia
Launching new julia process to run commands...
ERROR: LoadError: could not open file /home/runner/work/AcuteML.jl/AcuteML.jl/deps/SnoopCompile/precompile/precompile_AcuteML.jl
Stacktrace:
 [1] include at ./boot.jl:328 [inlined]
 [2] include_relative(::Module, ::String) at ./loading.jl:1105
 [3] include at ./Base.jl:31 [inlined]
 [4] include(::String) at /home/runner/work/AcuteML.jl/AcuteML.jl/src/AcuteML.jl:1
 [5] top-level scope at /home/runner/work/AcuteML.jl/AcuteML.jl/src/AcuteML.jl:1055
 [6] include at ./boot.jl:328 [inlined]
 [7] include_relative(::Module, ::String) at ./loading.jl:1105
 [8] include(::Module, ::String) at ./Base.jl:31
 [9] top-level scope at none:2
 [10] eval at ./boot.jl:330 [inlined]
 [11] eval(::Expr) at ./client.jl:425
 [12] top-level scope at ./none:3
in expression starting at /home/runner/work/AcuteML.jl/AcuteML.jl/src/AcuteML.jl:1055
ERROR: Failed to precompile AcuteML [6b84a0a3-a11a-41fa-a417-e600528de108] to /home/runner/.julia/compiled/v1.3/AcuteML/DX2uo_b3BsC.ji.
Stacktrace:
done.
 [1] error(::String) at ./error.jl:33
 [2] compilecache(::Base.PkgId, ::String) at ./loading.jl:1283
 [3] _require(::Base.PkgId) at ./loading.jl:1024
 [4] require(::Base.PkgId) at ./loading.jl:922
 [5] require(::Module, ::Symbol) at ./loading.jl:917
 [6] macro expansion at /home/runner/work/AcuteML.jl/AcuteML.jl/deps/SnoopCompile/snoopCompile.jl:13 [inlined]
```

**Comment by aminya** [comment]
Oh!! We cannot run SnoopCompile when the `__precompile__` is added to the source file! So, it should remove those lines of code before running the tests

**Comment by aminya** [comment]
I have written this but it doesn't work:

Consider we want to comment and uncomment this 
```julia
include("../deps/SnoopCompile/precompile/precompile_AcuteML.jl")
_precompile_()
```

```julia
packageName = "AcuteML"

function precompileDeactivator(packageName)

    file = open(joinpath(pwd(),"src","$packageName.jl"),"r+")

    packageText = read(file, String)

    packageEdited = foldl(replace,
                 (
                  "include(\"../deps/SnoopCompile/precompile/precompile_$packageName.jl\")" => "#include(\"../deps/SnoopCompile/precompile/precompile_$packageName.jl\")",
                  "_precompile_()" => "#_precompile_()",
                 ),
                 init = packageText)
                 print(packageEdited)
     write(file, packageEdited)
     close(file)
end

function precompileActivator(packageName)

    file = open(joinpath(pwd(),"src","$packageName.jl"),"r+")
    packageText = read(file, String)

    packageEdited = foldl(replace,
                 (
                  "#include(\"../deps/SnoopCompile/precompile/precompile_$packageName.jl\")" => "include(\"../deps/SnoopCompile/precompile/precompile_$packageName.jl\")",
                  "#_precompile_()" => "_precompile_()",
                 ),
                 init = packageText)

     print(file, packageEdited)
     close(file)
end


################################################################
precompileDeactivator(packageName);

rootPath = pwd()
cd(@__DIR__)

### Log the compiles
# This only needs to be run once (to generate log file)

SnoopCompile.@snoopc "$(pwd())/Snoop.log" begin

    # Use runtests.jl or your exmaples that uses package:

    using AcuteML, Pkg

    include(joinpath(dirname(dirname(pathof(AcuteML))), "test","runtests.jl"))
    include(joinpath(dirname(dirname(pathof(AcuteML))), "examples","extractor.jl"))
    include(joinpath(dirname(dirname(pathof(AcuteML))), "examples","constructor.jl"))
    include(joinpath(dirname(dirname(pathof(AcuteML))), "examples","templating","templating.jl"))


end

################################################################

### Parse the compiles and generate precompilation scripts
# This can be run repeatedly to tweak the scripts

data = SnoopCompile.read("$(pwd())/Snoop.log")

pc = SnoopCompile.parcel(reverse!(data[2]))
SnoopCompile.write("$(pwd())/precompile", pc)

################################################################
precompileActivator(packageName)
cd(rootPath)

```

**Comment by aminya** [comment]
I fixed the problem by updating the script. It works on my package now!
https://github.com/juliamatlab/MatLang/commit/4171d4630e618ff12487d65464e88b697b0abf0e/checks?check_suite_id=337510876

PR generated: https://github.com/juliamatlab/MatLang/pull/16

**Comment by KristofferC** [comment]
FWIW, I don't think generating precompile statements based on full test suites is a good idea in general, and should not be something that should be e.g. enabled by default on packages.

**Comment by timholy** [comment]
Yes, generally what makes sense for precompilation is quite different from what makes sense for testing. For example, for [FixedPointNumbers](https://github.com/JuliaMath/FixedPointNumbers.jl/blob/master/src/precompile.jl) and [ColorTypes](https://github.com/JuliaGraphics/ColorTypes.jl/blob/master/src/precompile.jl), I just wrote the precompile files by hand, with some guidance from SnoopCompile, because the tests exercise all kinds of weird types (e.g., `N3f5`) that are never used in practice.

**Comment by aminya** [comment]
One should add 3 things to a package to make the bot work:
worflow file
https://github.com/juliamatlab/MatLang/blob/master/.github/workflows/SnoopCompile.yml
the examples that calls the functions or Pkg.test()  https://github.com/juliamatlab/MatLang/blob/master/deps/SnoopCompile/snoopCompile.jl
Two lines of code that includes the precompile file
https://github.com/juliamatlab/MatLang/blob/0206d062096cf3cbe3b7c496d058319b04d8fed9/src/MatLang.jl#L27

Now I didn't choose test as the default but that's a script with good coverage of API

**Comment by timholy** [comment]
I think that a suitably-careful implementation makes this worthy of bot-ifying. What do you see as the path forward, @aminya?

**Comment by KristofferC** [comment]
I thought that https://github.com/timholy/SnoopCompile.jl/pull/49/commits/a238b735b82cb812f4d3a7bf5f9c7ff37e3f28e2 meant to use the tests as a default.

Does the snoop call really propagate into the new process started by `Pkg.test` or will this just record Pkg itself?

**Comment by timholy** [comment]
I don't think it will propagate.

**Comment by aminya** [comment]
> 
> 
> I think that a suitably-careful implementation makes this worthy of bot-ifying. What do you see as the path forward, @aminya?

Thanks!
Things we can improve are as follows:
- [ ] Creating a Github Action repository that has the workflow file.
- [ ] Do more testing for other packages to make sure everything is fine.
- [ ] We can do more automation, but it may conflict with packages that already have `__precompile__`, so I prefered to give an error to let the user edit the code  https://github.com/aminya/SnoopCompile.jl/blob/a238b735b82cb812f4d3a7bf5f9c7ff37e3f28e2/src/utilities.jl#L49

**Comment by aminya** [comment]
> 
> 
> I thought that [a238b73](https://github.com/timholy/SnoopCompile.jl/commit/a238b735b82cb812f4d3a7bf5f9c7ff37e3f28e2) meant to use the tests as a default.
> 
> Does the snoop call really propagate into the new process started by `Pkg.test` or will this just record Pkg itself?

It doesn't use that by default. My commit message is a little misleading. User can choose to use that method inside the call script (e.g. https://github.com/juliamatlab/MatLang/blob/master/deps/SnoopCompile/snoopCompile.jl)

**Comment by KristofferC** [comment]
> It doesn't use that by default. 

Okay, so the idea is that you give a "workload" and the bot uses SnoopCompile to make sure that the precompile statements are up to date based on that workload. That sounds better.

**Comment by timholy** [comment]
> There's a bit of a problem that different Julia versions and operating systems can have different "valid" precompile statements but might be fine in practice.

At least the most common reasons should be handled now automatically (if you don't know this already, which you probably do, see [NEWS.md](https://github.com/timholy/SnoopCompile.jl/blob/master/NEWS.md#version-11)), but I'm guessing you mean stuff like
```julia
if VERSION < v"1.3.0"
    foo(x, y) = 1
else
    foo(x) = 1
end
```
which of course is beyond what an automatic tool can handle. As you probably also know, it really comes down to functions rather than methods, since `precompile` itself just returns `true`/`false`:

```julia
julia> precompile(Tuple{typeof(+),String,String})
false

julia> "hi"+"Kristoffer"
ERROR: MethodError: no method matching +(::String, ::String)
Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...) at operators.jl:529
Stacktrace:
 [1] top-level scope at REPL[2]:1
```

**Comment by aminya** [comment]
I should add an option for custom precompile sentences. User should be able to point to custom precompile sentences, and `@snoopiBot` should merge those with automatically generated sentences.

**Comment by davidanthoff** [comment]
Could the bot run SnoopCompile on all different Julia versions and create different sets of precompile statements for different Julia versions? And the package then has include statements that are behind `@static` version checks? Seems weird, but if it is all automated, maybe not so crazy?

**Comment by aminya** [comment]
> Could the bot run SnoopCompile on all different Julia versions and create different sets of precompile statements for different Julia versions? And the package then has include statements that are behind `@static` version checks? Seems weird, but if it is all automated, maybe not so crazy?

It is possible to do that. I haven't found a way to automatically add code to the end of a module. That is why give error to the user to add the code themselves:
https://github.com/aminya/SnoopCompile.jl/blob/a238b735b82cb812f4d3a7bf5f9c7ff37e3f28e2/src/utilities.jl#L49

Is there a way to add two lines of code before a module end?

We can search in the package for `__precompile__` and if it is present give a warning of possible conflict with what the bot is adding.

Also my question: 
Doesn't that `@static` affect the precompiling behvaior?

**Comment by timholy** [comment]
> Is there a way to add two lines of code before a module end?

You could parse the file, look for such expressions, and delete them. Then `eval` the module.

**Comment by aminya** [comment]
Closed by #49
